### PR TITLE
Affine transform module - support DMA/CUDA via NPPI and HOST via open cv

### LINKED ISSUE #36: Affine Transform Module
Input is a Raw Image (Mono, YUV or RGB) 
Output is also a Raw Image (Mono, YUV or RGB) 

Affine matrix is provided as a dynamic property: 
https://docs.opencv.org/3.4.15/d4/d61/tutorial_warp_affine.html

We should try support at least two Interpolation methods: Bilinear: (INTER_LINEAR in openCV) and BiCubic: (INTER_CUBIC in OpenCV)

### PR DESCRIPTION
**Pull request by venkat0907**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #36

**Description**
Finished NPPI based affine transform module.
Supports YUV444, *RGB, MONO.
Added tests and data.

**Alternative(s) considered**

Have you considered any alternatives? And if so, why have you chosen the approach in this PR?
Alternative was to implement in opencv to support CPU - which is yet to be done.

**Type**

Type Choose one:  Feature

**Screenshots (if applicable)**

**Checklist**

- [x] I have read the [Contribution Guidelines](https://github.com/Apra-Labs/ApraPipes/wiki/Contribution-Guidelines)
- [x] I have written Unit Tests
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach

### CODE REVIEW

File: base/test/affinetransform_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+	f.affineTransform->step();
+	f.copy2->step();
+
+	auto outputPinId = f.copy2->getAllOutputPinsByType(FrameMetadata::RAW_IMAGE_PLANAR)[0];
+	auto frames = f.m3->pop();
+	BOOST_TEST((frames.find(outputPinId) != frames.end()));
+	auto outFrame = frames[outputPinId];
+	BOOST_TEST(outFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE_PLANAR);
+	Test_Utils::saveOrCompare("./data/testOutput/affinetransform_tests_yuv444_1920x1080.raw", (const uint8_t*)outFrame->data(), outFrame->size(), 0);
+}
```
Comment (mraduldubey) [inline base/test/affinetransform_tests.cpp:388]: add one test for dynamic change of properties aka getProps and setProps

**Review by mraduldubey — CHANGES_REQUESTED** [review]
Summary:
1. Let us support HOST memory via opencv
2. Add tests for HOST memory 
3. Add test for dynamic property change
4. Provide some clarifications/comments on interpolation

File: base/test/affinetransform_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+		BOOST_TEST(copy1->init());
+		BOOST_TEST(affineTransform->init());
+		BOOST_TEST(copy2->init());
+		BOOST_TEST(m3->init());
+	}
+	boost::shared_ptr<FileReaderModule>fileReader;
+	boost::shared_ptr<Module>copy1;
+	boost::shared_ptr<Module>affineTransform;
+	boost::shared_ptr<Module>copy2;
+	boost::shared_ptr<ExternalSinkModule>m3;
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/test/affinetransform_tests.cpp:70]: Resolved(changed it to sink)

File: base/src/AffineTransform.cpp [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+bool AffineTransform::term()
+{
+	mDetail.reset();
+	return Module::term();
+}
+
+bool AffineTransform::process(frame_container &frames)
+{
+	mDetail->InputFrame = frames.cbegin()->second;
+	mDetail->OutputFrame = makeFrame(mDetail->mFrameLength);
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/src/AffineTransform.cpp:447]: Resolved(changed it to mOutputFrameLength)

File: base/src/AffineTransform.cpp [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+		mDetail.reset(new DeatilCUDA(mProp));
+	}
+
+	else if (memType = FrameMetadata::MemType::DMABUF)
+	{
+		mDetail.reset(new DetailDMA(mProp));
+	}
+
+	else
+	{
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/src/AffineTransform.cpp:482]: Resolved

File: base/src/AffineTransform.cpp [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+	size_t mFrameLength;
+	frame_sp InputFrame;
+	frame_sp OutputFrame;
+	void* OutputPtr;
+	void* InputPtr;
+	std::string mOutputPinId;
+	framemetadata_sp mOutputMetadata;
+	AffineTransformProps props;
+
+	bool setMetadataHelper(framemetadata_sp &input, framemetadata_sp &output)
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/src/AffineTransform.cpp:293]: Resolved(made it protected)

File: base/test/affinetransform_tests.cpp [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+	f.copy1->step();
+	f.affineTransform->step();
+	f.copy2->step();
+
+	auto outputPinId = f.copy2->getAllOutputPinsByType(FrameMetadata::RAW_IMAGE)[0];
+	auto frames = f.m3->pop();
+	BOOST_TEST((frames.find(outputPinId) != frames.end()));
+	auto outFrame = frames[outputPinId];
+	BOOST_TEST(outFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE);
+	Test_Utils::saveOrCompare("./data/testOutput/affinetransform_tests_frame_400x400_rgba.raw", (const uint8_t*)outFrame->data(), outFrame->size(), 0);
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/test/affinetransform_tests.cpp:356]: added

File: base/test/affinetransform_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+	f.copy1->step();
+	f.affineTransform->step();
+	f.copy2->step();
+
+	auto outputPinId = f.copy2->getAllOutputPinsByType(FrameMetadata::RAW_IMAGE)[0];
+	auto frames = f.m3->pop();
+	BOOST_TEST((frames.find(outputPinId) != frames.end()));
+	auto outFrame = frames[outputPinId];
+	BOOST_TEST(outFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE);
+	Test_Utils::saveOrCompare("./data/testOutput/affinetransform_tests_frame_all_400x400_rgba.raw", (const uint8_t*)outFrame->data(), outFrame->size(), 0);
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/test/affinetransform_tests.cpp:376]: added

File: base/test/affinetransform_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+	f.copy1->step();
+	f.affineTransform->step();
+	f.copy2->step();
+
+	auto outputPinId = f.copy2->getAllOutputPinsByType(FrameMetadata::RAW_IMAGE_PLANAR)[0];
+	auto frames = f.m3->pop();
+	BOOST_TEST((frames.find(outputPinId) != frames.end()));
+	auto outFrame = frames[outputPinId];
+	BOOST_TEST(outFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE_PLANAR);
+	Test_Utils::saveOrCompare("./data/testOutput/affinetransform_tests_yuv444_1920x1080.raw", (const uint8_t*)outFrame->data(), outFrame->size(), 0);
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/test/affinetransform_tests.cpp:435]: added

File: base/test/affinetransform_tests.cpp [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+	f.copy1->step();
+	f.affineTransform->step();
+	f.copy2->step();
+
+	auto outputPinId = f.copy2->getAllOutputPinsByType(FrameMetadata::RAW_IMAGE)[0];
+	auto frames = f.m3->pop();
+	BOOST_TEST((frames.find(outputPinId) != frames.end()));
+	auto outFrame = frames[outputPinId];
+	BOOST_TEST(outFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE);
+	Test_Utils::saveOrCompare("./data/testOutput/affinetransform_tests_8bit_frame_1280x720_bgra.raw", (const uint8_t*)outFrame->data(), outFrame->size(), 0);
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/test/affinetransform_tests.cpp:396]: added

File: base/test/affinetransform_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+	f.copy1->step();
+	f.affineTransform->step();
+	f.copy2->step();
+
+	auto outputPinId = f.copy2->getAllOutputPinsByType(FrameMetadata::RAW_IMAGE)[0];
+	auto frames = f.m3->pop();
+	BOOST_TEST((frames.find(outputPinId) != frames.end()));
+	auto outFrame = frames[outputPinId];
+	BOOST_TEST(outFrame->getMetadata()->getFrameType() == FrameMetadata::RAW_IMAGE);
+	Test_Utils::saveOrCompare("./data/testOutput/affinetransform_tests-RGB_scale_rotate_frame_1280x720_rgb.raw", (const uint8_t*)outFrame->data(), outFrame->size(), 0);
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/test/affinetransform_tests.cpp:295]: added

File: base/src/AffineTransform.cpp [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+	framemetadata_sp metadata = getFirstInputMetadata();
+	FrameMetadata::FrameType frameType = metadata->getFrameType();
+	if (frameType != FrameMetadata::RAW_IMAGE && frameType != FrameMetadata::RAW_IMAGE_PLANAR)
+	{
+		LOG_ERROR << "<" << getId() << ">::validateInputPins input frameType is expected to be RAW_IMAGE or RAW_IMAGE_PLANAR. Actual<" << frameType << ">";
+		return false;
+	}
+
+	FrameMetadata::MemType memType = metadata->getMemType();
+	if (memType != FrameMetadata::MemType::CUDA_DEVICE && memType != FrameMetadata::MemType::DMABUF)
```
[the remark this answers is not in the document]
Comment (venkat0907) [reply base/src/AffineTransform.cpp]: added host support

File: base/test/affinetransform_tests.cpp (hunk 1 above)
Comment (venkat0907) [reply base/test/affinetransform_tests.cpp:464]: added

**Review by mraduldubey — DISMISSED** [review]

**Review by mraduldubey — APPROVED** [review]

**Pull request merged**